### PR TITLE
Switch to esnext imports

### LINKED ISSUE #5751: Jest tests failing when importing ES modules
It looks like our current jest config doesn't support ES modules (because it's running in a node.js environment). The blueprint packages provide ES modules by default. We can import the commonjs modules if need be (e.g. `import { Button } from '@blueprintjs/core/lib/cjs/components/button/buttons'`) but it seems like we will run into this issue more as ES modules becomes the standard module type.

Output from [Travis build](https://travis-ci.org/jupyterlab/jupyterlab/jobs/466235657):

```
@jupyterlab/test-application: $ python run.py --coverage
@jupyterlab/test-application: [I 22:32:48.911 JestApp] Writing notebook server cookie secret to /tmp/tmpeywu3yvy/jupyter_runtime/notebook_cookie_secret
@jupyterlab/test-application: [I 22:32:48.938 JestApp] Serving notebooks from local directory: /tmp/mock_contentsp7r4iwwc
@jupyterlab/test-application: [I 22:32:48.938 JestApp] The Jupyter Notebook is running at:
@jupyterlab/test-application: [I 22:32:48.938 JestApp] http://localhost:8888/?token=d05099444bb13d70c8fce4c8c799ca90d83774cba2ffc0d8
@jupyterlab/test-application: [I 22:32:48.938 JestApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).
@jupyterlab/test-application: [C 22:32:48.938 JestApp] 
@jupyterlab/test-application:     
@jupyterlab/test-application:     Copy/paste this URL into your browser when you connect for the first time,
@jupyterlab/test-application:     to login with a token:
@jupyterlab/test-application:         http://localhost:8888/?token=d05099444bb13d70c8fce4c8c799ca90d83774cba2ffc0d8
@jupyterlab/test-application: [I 22:32:48.941 JestApp] > node /home/travis/build/jupyterlab/jupyterlab/node_modules/jest/bin/jest.js --coverage --runInBand
@jupyterlab/test-application: jest-haste-map: @providesModule naming collision:
@jupyterlab/test-application:   Duplicate module name: @jupyterlab/application-top
@jupyterlab/test-application:   Paths: /home/travis/build/jupyterlab/jupyterlab/dev_mode/package.json collides with /home/travis/build/jupyterlab/jupyterlab/jupyterlab/staging/package.json
@jupyterlab/test-application: This warning is caused by a @providesModule declaration with the same name across two different files.
@jupyterlab/test-application:  FAIL  src/shell.spec.ts
@jupyterlab/test-application:   ● Test suite failed to run
@jupyterlab/test-application:     Jest encountered an unexpected token
@jupyterlab/test-application:     This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.
@jupyterlab/test-application:     By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".
@jupyterlab/test-application:     Here's what you can do:
@jupyterlab/test-application:      • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
@jupyterlab/test-application:      • If you need a custom transformation specify a "transform" option in your config.
@jupyterlab/test-application:      • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.
@jupyterlab/test-application:     You'll find more details and examples of these config options in the docs:
@jupyterlab/test-application:     https://jestjs.io/docs/en/configuration.html
@jupyterlab/test-application:     Details:
@jupyterlab/test-application:     /home/travis/build/jupyterlab/jupyterlab/node_modules/@blueprintjs/core/lib/esm/components/button/buttons.js:6
@jupyterlab/test-application:     import * as tslib_1 from "tslib";
@jupyterlab/test-application:     ^^^^^^
@jupyterlab/test-application:     SyntaxError: Unexpected token import
@jupyterlab/test-application:       11 | Object.defineProperty(exports, "__esModule", { value: true });
@jupyterlab/test-application:       12 | const React = __importStar(require("react"));
@jupyterlab/test-application:     > 13 | const buttons_1 = require("@blueprintjs/core/lib/esm/components/button/buttons");
@jupyterlab/test-application:          |                   ^
@jupyterlab/test-application:       14 | const icon_1 = require("@blueprintjs/core/lib/esm/components/icon/icon");
@jupyterlab/test-application:       15 | const collapse_1 = require("@blueprintjs/core/lib/esm/components/collapse/collapse");
@jupyterlab/test-application:       16 | const inputGroup_1 = require("@blueprintjs/core/lib/esm/components/forms/inputGroup");
@jupyterlab/test-application:       at ScriptTransformer._transformAndBuildScript (node_modules/jest-runtime/build/script_transformer.js:403:17)
@jupyterlab/test-application:       at Object.<anonymous> (packages/ui-components/lib/index.js:13:19)
@jupyterlab/test-application:       at Object.<anonymous> (packages/apputils/lib/toolbar.js:13:25)
```

### PR DESCRIPTION
**Pull request by blink1073**
Fixes #5751.  We are now using ts-jest@24.0.2, which has support for es6 modules.  This PR switches to using `import()` syntax and I also verified that the following works:

```diff
diff --git a/packages/application/src/index.ts b/packages/application/src/index.ts
index 22316885a..e30fd725c 100644
--- a/packages/application/src/index.ts
+++ b/packages/application/src/index.ts
@@ -17,3 +17,7 @@ export { IRouter, Router } from './router';
 export { ILabShell, LabShell } from './shell';
 
 export { ILabStatus } from './status';
+
+import { Button } from '@blueprintjs/core/lib/cjs/components/button/buttons';
+
+console.log(Button);
```

I also ran a build and made sure that the terminal launches (which uses `import()`):

image

### CODE REVIEW

**Comment by blink1073** [comment]
cc @gnestor

**Comment by blink1073** [comment]
cc @bollwyvl

**Comment by jasongrout** [comment]
What browsers support the esnext target?

**Comment by blink1073** [comment]
This isn't intended to be used directly in the browser.  Webpack coverts this to use an ensure.

**Comment by jasongrout** [comment]
> This isn't intended to be used directly in the browser. Webpack coverts this to use an ensure.

I'm not talking about the imports. IIRC, the TS esnext target does other things like passing through async/await, etc. I want to be careful about how this is going to affect our minimum browser version.

File: packages/vega5-extension/src/index.ts [hunk 1]
```diff
@@ -188,22 +188,7 @@ namespace Private {
...
-          resolve(vega);
-        },
-        (error: any) => {
-          console.error(error);
-          reject();
-        },
-        'vega'
-      );
-    });
+    vegaReady = import('vega-embed');
```
Comment (bollwyvl) [inline packages/vega5-extension/src/index.ts]: :raised_hands:

**Comment by bollwyvl** [comment]
Fantastic stuff!

> This isn't intended to be used directly in the browser. Webpack coverts this to use an ensure.

Yet. But `pika/web` etc. have the potential to remove the end user requirement for node/npm in order to add labextensions, make downstream packagers far happier, and more benefits than i can name. Not sure if they can easily handle the duplication munging that lead us to WP. Of course, this ain't free: we'd have to start pretty deep in the stack to get to a full `web_modules` deploy. `phosphor` should be pretty easy. Long poles include `react`, `ajv` and `codemirror`. 

The real tale of the tape will be the response time of a representative deployment (e.g. binder) to an reasonable bandwidth (Fast 3G) consumer. I'm going to wager `8888` users would barely notice.

>  I want to be careful about how this is going to affect our minimum browser version.

Not just that, but writing stuff that works in node requires emitting `mjs`... but again, you can just have another target.

Keeping a bundled legacy build around as an app of last resort would be fine, but we never really supported ie/edge (neither will MS anymore) or mobile, so i don't know which market share we're leaving out: Firefox ESR is good to go. Back at the Company I Shall Not Name, I actually maintained such a compatibility build for a while... _another_ 100mb of node stuff at build time for css and a couple megs of polyfills to make all the bells and whistles work. Not fun.

File: packages/terminal-extension/src/index.ts [hunk 2]
```diff
@@ -403,19 +403,7 @@ namespace Private {
...
-        // this argument MUST be named `require` for the WebPack parser
-        require => resolve(require('@jupyterlab/terminal/lib/widget')),
-        (error: any) => {
-          showErrorMessage(error);
-          reject();
-        },
-        'terminal'
-      );
-    });
+    widgetReady = import('@jupyterlab/terminal/lib/widget');
```
Comment (bollwyvl) [inline packages/terminal-extension/src/index.ts]: dagnabit, just got that working :crying_cat_face:

**Comment by blink1073** [comment]
The transpiled code uses the `_awaiter` stub, not the `await` keyword.

**Comment by jasongrout** [comment]
I was getting confused. There are two places in tsconfig that you can have `esnext`. You changed the modules to be `esnext`, and I was talking about changing the target to be `esnext`. Yes, +1 on changing modules to `esnext` if it makes things easier.

This does mean that our original mantra of "jlab extensions are commonjs modules" is obsolete, and can now be "jlab extensions are commonjs or esnext modules".

**Comment by jasongrout** [comment]
I've made switching to es2017 ts *target* https://github.com/jupyterlab/jupyterlab/pull/6224

**Comment by jasongrout** [comment]
The last test failures look related to the changes in this PR.

File: testutils/tsconfig.json [hunk 3]
```diff
@@ -1,8 +1,22 @@
 {
-  "extends": "../tsconfigbase",
   "compilerOptions": {
     "outDir": "lib",
     "rootDir": "src",
+    "allowSyntheticDefaultImports": true,
```
Comment (jasongrout) [inline testutils/tsconfig.json]: I think you can still inherit, and just override the module key.

Comment (jasongrout) [reply testutils/tsconfig.json]: which, of course, you already tried. Not sure why it didn't work...

**Comment by jasongrout** [comment]
I switched back to inheriting from the base tsconfig, and the build and test build seemed to work fine for me. I also merged in master to get the es2017 target change.

**Comment by blink1073** [comment]
Note: I spent some time with this this morning, but I could not figure out how to get babel to process our files in lib.  I was looking at https://kulshekhar.github.io/ts-jest/user/config/babelConfig and https://kulshekhar.github.io/ts-jest/user/config/#jest-preset.

**Comment by jasongrout** [comment]
Some relevant issues:

https://github.com/facebook/jest/issues/4842

When I run `jlpm run test`, I get three errors:

This, which seems that it is an issue with the jlab library it is importing

```
@jupyterlab/test-application: FAIL src/shell.spec.ts
@jupyterlab/test-application:   ● Test suite failed to run
@jupyterlab/test-application:     Jest encountered an unexpected token
@jupyterlab/test-application:     This usually means that you are trying to import a file which Jest cannot parse, e.g. it's not plain JavaScript.
@jupyterlab/test-application:     By default, if Jest sees a Babel config, it will use that to transform your files, ignoring "node_modules".
@jupyterlab/test-application:     Here's what you can do:
@jupyterlab/test-application:      • To have some of your "node_modules" files transformed, you can specify a custom "transformIgnorePatterns" in your config.
@jupyterlab/test-application:      • If you need a custom transformation specify a "transform" option in your config.
@jupyterlab/test-application:      • If you simply want to mock your non-JS modules (e.g. binary assets) you can stub them out with the "moduleNameMapper" config option.
@jupyterlab/test-application:     You'll find more details and examples of these config options in the docs:
@jupyterlab/test-application:     https://jestjs.io/docs/en/configuration.html
@jupyterlab/test-application:     Details:
@jupyterlab/test-application:     /[snip]/jupyterlab/packages/services/lib/index.js:3
@jupyterlab/test-application:     export * from './config';
@jupyterlab/test-application:     ^^^^^^
@jupyterlab/test-application:     SyntaxError: Unexpected token export
@jupyterlab/test-application:       4 | Object.defineProperty(exports, "__esModule", { value: true });
@jupyterlab/test-application:       5 | const simulate_event_1 = require("simulate-event");
@jupyterlab/test-application:     > 6 | const services_1 = require("@jupyterlab/services");
@jupyterlab/test-application:         |                    ^
@jupyterlab/test-application:       7 | const apputils_1 = require("@jupyterlab/apputils");
@jupyterlab/test-application:       8 | const coreutils_1 = require("@phosphor/coreutils");
@jupyterlab/test-application:       9 | const signaling_1 = require("@phosphor/signaling");
@jupyterlab/test-application:       at ScriptTransformer._transformAndBuildScript (node_modules/@jest/transform/build/ScriptTransformer.js:471:17)
@jupyterlab/test-application:       at ScriptTransformer.transform (node_modules/@jest/transform/build/ScriptTransformer.js:513:25)
@jupyterlab/test-application:       at Object.<anonymous> (testutils/lib/index.js:6:20)
```

and this, which seems it is an issue with importing a .css file is not supported in jest:

```
@jupyterlab/test-application: FAIL src/router.spec.ts
@jupyterlab/test-application:   ● Test suite failed to run
@jupyterlab/test-application:     /[snip]/jupyterlab/packages/application/lib/index.js:4
@jupyterlab/test-application:     import '../style/index.css';
@jupyterlab/test-application:            ^^^^^^^^^^^^^^^^^^^^
@jupyterlab/test-application:     SyntaxError: Unexpected string
@jupyterlab/test-application:       4 | import { expect } from 'chai';
@jupyterlab/test-application:       5 | 
@jupyterlab/test-application:     > 6 | import { Router } from '@jupyterlab/application';
@jupyterlab/test-application:         | ^
@jupyterlab/test-application:       7 | 
@jupyterlab/test-application:       8 | import { CommandRegistry } from '@phosphor/commands';
@jupyterlab/test-application:       9 | 
@jupyterlab/test-application:       at ScriptTransformer._transformAndBuildScript (node_modules/@jest/transform/build/ScriptTransformer.js:471:17)
@jupyterlab/test-application:       at ScriptTransformer.transform (node_modules/@jest/transform/build/ScriptTransformer.js:513:25)
@jupyterlab/test-application:       at Object.<anonymous> (tests/test-application/src/router.spec.ts:6:1)
```

**Comment by jasongrout** [comment]
note to self: this looks like a relevant issue: https://github.com/facebook/jest/issues/4842

**Comment by blink1073** [comment]
Found the issue: `babel-jest@24` doesn't respect the inputs given by `ts-jest` or a `.babelrc` file, only a `babel.config.js` file.  Babel is quite a moving target these days...

**Comment by blink1073** [comment]
I fixed `test-application` in place, need to make it uniform across the jest tests.

**Comment by jasongrout** [comment]
Will this also take care of the issues with the jest coverage runs returning errors?

**Comment by jasongrout** [comment]
Does this mean we can also switch the test tsconfig back to `module: esnext`, or do we still need to keep it `module: commonjs`?

**Comment by jasongrout** [comment]
> Does this mean we can also switch the test tsconfig back to `module: esnext`, or do we still need to keep it `module: commonjs`?

(I guess it's testutils tsconfig, not the test tsconfigs...)

**Comment by blink1073** [comment]
Right, that is for the testutils package itself, because it uses `require()`.  This will address the syntax errors we saw with the `import` and `export`.

**Comment by gnestor** [comment]
@bollwyvl Have you taken a stab at creating a `web_modules` for jupyterlab? I'm curious what sort of issues we would run into. 

I did a bit of research a while ago and I was under the impression that rollup outputs stuff as ES6 modules but I couldn't really find any examples of people using rollup to build apps this way, just libraries. @pikapkg/web is clearly built for that purpose so I'm very excited to start playing with it.

**Comment by blink1073** [comment]
Yikes, this was the missing piece: https://github.com/webpack-contrib/istanbul-instrumenter-loader#with-babel

**Comment by blink1073** [comment]
image

**Comment by vidartf** [comment]
The last commit here switches coreutils and services back to `commonjs`. What was the rationale for this? Just for reference.

**Comment by blink1073** [comment]
They are both meant to be consumed from node clients.

**Pull request merged**